### ISSUE TITLE
Can't call rollback when autocommit=true

### ISSUE THREAD

**Issue by guy93r** [brettwooldridge/HikariCP#1195]
#### Environment
```
HikariCP version: 2.3.13
JDK version     : 1.8.0_111
Database        : MySQL
Driver version  : 5.1.34
```

hey!
while running with mybatis+HikariCP I encounter this warning message:
```
WARN [main] (ConnectionProxy.java:113) - Connection com.mysql.jdbc.JDBC4Connection@6482eef (HikariPool-0) marked as broken because of SQLSTATE(08003), ErrorCode(0).
com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Can't call rollback when autocommit=true
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:377)
	at com.mysql.jdbc.Util.getInstance(Util.java:360)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:956)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:935)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:924)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:870)
	at com.mysql.jdbc.ConnectionImpl.rollback(ConnectionImpl.java:4590)
	at com.zaxxer.hikari.proxy.ConnectionProxy.close(ConnectionProxy.java:213)
	at org.apache.ibatis.transaction.jdbc.JdbcTransaction.close(JdbcTransaction.java:92)
	at org.apache.ibatis.executor.BaseExecutor.close(BaseExecutor.java:90)
	at org.apache.ibatis.executor.CachingExecutor.close(CachingExecutor.java:64)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.close(DefaultSqlSession.java:264)
	at org.apache.ibatis.session.SqlSessionManager$SqlSessionInterceptor.invoke(SqlSessionManager.java:364)
	at com.sun.proxy.$Proxy46.selectList(Unknown Source)
	at org.apache.ibatis.session.SqlSessionManager.selectList(SqlSessionManager.java:206)
	at org.apache.ibatis.binding.MapperMethod.executeForMany(MapperMethod.java:137)
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:75)
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:59)
	at com.sun.proxy.$Proxy47.getGroups(Unknown Source)
```

with loglevel=DEBUG i get this:
```
DEBUG [main] (JdbcTransaction.java:122) - Resetting autocommit to true on JDBC Connection [HikariConnectionProxy(1313459755) wrapping com.mysql.jdbc.JDBC4Connection@6482eef]
DEBUG [timeoutChecker-0] (JdbcTransaction.java:122) - Resetting autocommit to true on JDBC Connection [HikariConnectionProxy(619534860) wrapping com.mysql.jdbc.JDBC4Connection@68c022c]
DEBUG [timeoutChecker-0] (JdbcTransaction.java:90) - Closing JDBC Connection [HikariConnectionProxy(619534860) wrapping com.mysql.jdbc.JDBC4Connection@68c022c]
DEBUG [main] (JdbcTransaction.java:90) - Closing JDBC Connection [HikariConnectionProxy(1313459755) wrapping com.mysql.jdbc.JDBC4Connection@6482eef]
DEBUG [timeoutChecker-0] (ConnectionProxy.java:211) - HikariPool-0 Performing rollback on com.mysql.jdbc.JDBC4Connection@68c022c due to dirty commit state.
DEBUG [main] (ConnectionProxy.java:211) - HikariPool-0 Performing rollback on com.mysql.jdbc.JDBC4Connection@6482eef due to dirty commit state.
```
it's a warning message I get once every few seconds and its only printed as a warning.
does someone know why this is happening or how can I fix it?
thanks!

**Comment by amanullah92** [comment]
What's your default autocommit value?
When the connection is returned to pool, it is reset to default value, which i think in your case is false. But autocommit=false might mean an open transaction, so Hikari prefers to do a rollback just to be safe.

For the first issue, it seems your query is being terminated in a non-standard manner; can you trace what the application is doing exactly? Is a transaction being rolled back because of an error? Is the rollback being trigger by code or internally by ORM? 

In any case, it usually not an issue with Hikari; we've seen this behavior with Eclipselink ORM also where it tries a rollback on fresh connection when original transaction fails because it could not acquire a connection! Totally wastes a connection..

**Comment by guy93r** [comment]
hey, thanks for the response! 
I'm using the default auto-commit setting(=true).
also, I think there is only one issue and both of the logs are from the same file/time. 
there is no error, I can't even catch the exception as it's handled internally, the transaction is managed by mybatis. 
so are you saying it's a mybatis problem?

thanks!

**Comment by brettwooldridge** [comment]
@guy93r HikariCP is logging this...
```
Resetting autocommit to true on JDBC Connection [HikariConnectionProxy(1313459755) wrapping com.mysql.jdbc.JDBC4Connection@6482eef]
```
Because your pool is configured as ``autoCommit=true``, but either your code, or mybatis, is changing the connection to ``autoCommit=false`` and then returning it to the pool.

I would recommend setting ``autoCommit=false`` to match what your code/mybatis is doing anyway.  Resetting ``autoCommit`` at the close of every connection for many databases is an additional roundtrip over the wire, which is unnecessarily expensive.

**Comment by smallufo** [comment]
I am facing the same problem. (found this thread by searching keywords).
My code works well with SpringBoot 2.0.4.RELEASE
When hibernate.version = 5.2.9.Final  , it works fine.
But when hibernate.version = 5.2.10.Final , it throws exception  :  
(It's running a unit test , with truly hit to db )

```
2018-08-14 03:34:31.276 [main] WARN  c.z.h.p.ProxyConnection - HikariPool-1 - Connection com.mysql.jdbc.JDBC4Connection@53a645a2 marked as broken because of SQLSTATE(08003), ErrorCode(0)
com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Can't call rollback when autocommit=true
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
	at com.mysql.jdbc.Util.getInstance(Util.java:387)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:917)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:896)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:885)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:860)
	at com.mysql.jdbc.ConnectionImpl.rollback(ConnectionImpl.java:4618)
	at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:370)
	at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java)
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.rollback(AbstractLogicalConnectionImplementor.java:116)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.rollback(JdbcResourceLocalTransactionCoordinatorImpl.java:239)
	at org.hibernate.engine.transaction.internal.TransactionImpl.rollback(TransactionImpl.java:100)
	at org.springframework.orm.jpa.JpaTransactionManager.doRollback(JpaTransactionManager.java:563)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processRollback(AbstractPlatformTransactionManager.java:838)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.rollback(AbstractPlatformTransactionManager.java:812)
	at org.springframework.test.context.transaction.TransactionContext.endTransaction(TransactionContext.java:128)
	at org.springframework.test.context.transaction.TransactionalTestExecutionListener.afterTestMethod(TransactionalTestExecutionListener.java:216)
	at org.springframework.test.context.TestContextManager.afterTestMethod(TestContextManager.java:443)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:94)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

org.springframework.transaction.TransactionSystemException: Could not roll back JPA transaction; nested exception is org.hibernate.TransactionException: Unable to rollback against JDBC Connection

	at org.springframework.orm.jpa.JpaTransactionManager.doRollback(JpaTransactionManager.java:567)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processRollback(AbstractPlatformTransactionManager.java:838)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.rollback(AbstractPlatformTransactionManager.java:812)
	at org.springframework.test.context.transaction.TransactionContext.endTransaction(TransactionContext.java:128)
	at org.springframework.test.context.transaction.TransactionalTestExecutionListener.afterTestMethod(TransactionalTestExecutionListener.java:216)
	at org.springframework.test.context.TestContextManager.afterTestMethod(TestContextManager.java:443)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:94)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.hibernate.TransactionException: Unable to rollback against JDBC Connection
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.rollback(AbstractLogicalConnectionImplementor.java:121)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.rollback(JdbcResourceLocalTransactionCoordinatorImpl.java:239)
	at org.hibernate.engine.transaction.internal.TransactionImpl.rollback(TransactionImpl.java:100)
	at org.springframework.orm.jpa.JpaTransactionManager.doRollback(JpaTransactionManager.java:563)
	... 24 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Can't call rollback when autocommit=true
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
	at com.mysql.jdbc.Util.getInstance(Util.java:387)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:917)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:896)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:885)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:860)
	at com.mysql.jdbc.ConnectionImpl.rollback(ConnectionImpl.java:4618)
	at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:370)
	at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java)
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.rollback(AbstractLogicalConnectionImplementor.java:116)
	... 27 more

2018-08-14 03:34:31.300 [Thread-2] INFO  c.z.h.HikariDataSource - HikariPool-2 - Shutdown initiated...
2018-08-14 03:34:31.308 [Thread-2] INFO  c.z.h.HikariDataSource - HikariPool-2 - Shutdown completed.

```

My persistence.xml (transaction-type="RESOURCE_LOCAL")  has this line :  
```<property name="hibernate.connection.provider_disables_autocommit" value="true"/>```

I added ```spring.datasource.hikari.auto-commit=false``` to spring's ```application.properties``` , but still no help.

other environments FYI : 
```
[INFO] |  +- org.hibernate:hibernate-hikaricp:jar:5.2.10.Final:compile
[INFO] |  |  \- com.zaxxer:HikariCP:jar:2.7.9:compile
```

and this latest (3.2.0) , still no help
```
[INFO] |  +- org.hibernate:hibernate-hikaricp:jar:5.2.10.Final:compile
[INFO] |  |  \- com.zaxxer:HikariCP:jar:3.2.0:compile
```


TL;DR  :  
hibernate 5.2.9 + HikariCP (3.2.0 or 2.7.9) :  OK
hibernate 5.2.10 + HikariCP (3.2.0 or 2.7.9) :  FAILED

mysql is same : mysql:mysql-connector-java:jar:5.1.46:compile
jdk1.8.0_144.jdk  , OSX

**Comment by smallufo** [comment]
According to hibernate 5.2.10's changelog : 
https://github.com/hibernate/hibernate-orm/blob/master/changelog.txt#L722

It seems [HHH-11542] is the problem here
https://hibernate.atlassian.net/browse/HHH-11542

I added 
```<property name="hibernate.connection.autocommit" value="false"/>``` to persistence.xml
And it works well now.
hibernate 5.2.17 + HikariCP:jar:3.2.0 , OK now.

Well , in fact , I am not sure the difference between ```hibernate.connection.autocommit=false``` and ```hibernate.connection.provider_disables_autocommit=true```. But , anyway , it works now.

And it seems I can delete ```hibernate.connection.provider_disables_autocommit=true``` line , and it seems working well (for now).
